pinot-broker/src/main/java/org/apache/pinot/broker/routing/timeboundary/TimeBoundaryManager.java

```diff
@@ -142,17 +153,42 @@ private long extractEndTimeMsFromSegmentZKMetadataZNRecord(String segment, @Null
     return endTimeMs;
   }
 
-  private void updateTimeBoundaryInfo(long maxEndTimeMs) {
-    if (maxEndTimeMs > 0) {
-      String timeBoundary = _timeFormatSpec.fromMillisToFormat(maxEndTimeMs - _timeOffsetMs);
-      TimeBoundaryInfo currentTimeBoundaryInfo = _timeBoundaryInfo;
-      if (currentTimeBoundaryInfo == null || !currentTimeBoundaryInfo.getTimeValue().equals(timeBoundary)) {
-        _timeBoundaryInfo = new TimeBoundaryInfo(_timeColumn, timeBoundary);
-        LOGGER.info("Updated time boundary to: {} for table: {}", timeBoundary, _offlineTableName);
+  private synchronized void updateTimeBoundaryInfo(Long enforcedTimeBoundary, long maxEndTimeMs,
+      boolean idealStateReffered) {
+    TimeBoundaryInfo currentTimeBoundaryInfo = _timeBoundaryInfo;
+    Long finalTimeBoundaryMs = null;
+    boolean isEnforced = false;
+    boolean validTimeBoundaryFound = false;
+
+    if (enforcedTimeBoundary != null) {
+      finalTimeBoundaryMs = enforcedTimeBoundary;
+      isEnforced = true;
+      validTimeBoundaryFound = true;
+      LOGGER.info("Enforced table time boundary in use: {} for table: {}", enforcedTimeBoundary, _offlineTableName);
+    } else if (idealStateReffered || !currentTimeBoundaryInfo.isEnforced()) {
+      if (maxEndTimeMs > 0) {
+        finalTimeBoundaryMs = maxEndTimeMs - _timeOffsetMs;
+        validTimeBoundaryFound = true;
+      } else {
+        LOGGER.warn("Failed to find segment with valid end time for table: {}, no time boundary generated",
+            _offlineTableName);
+      }
+    } else {
+      validTimeBoundaryFound = true;
+      LOGGER.info("Skipping time boundary update since enforced time boundary exists");
+    }
+
+    if (validTimeBoundaryFound) {
+      if (finalTimeBoundaryMs != null) {
```

Comment:
We always want to update the metrics. Currently when user refresh a segment, the metric won't be updated.

Reply:
Good catch! Ack